### PR TITLE
[ATen][ProxyTensor] Preserve unbacked batch dims in SDPA tracing

### LINKED ISSUE #3324: [graph_trainer] SDPA tracing fails with unbacked dynamic batch dimension
## Summary

TorchTitan graph trainer chunk-pass tests cannot currently trace a real DeepSeek transformer block with `ScaledDotProductAttention` when the batch dimension is marked as an unbacked dynamic symbol. The trace fails before TorchTitan's chunk pass runs because SDPA's fake/proxy path asks for a concrete `aten.size.default` result on a tensor whose batch dimension is unbacked.

This blocks testing `chunk_batch` on a real attention-containing transformer block without replacing the inner attention module in the test.

## Minimal repro

```python
import torch
import torch.nn.functional as F
from torch._dynamo.source import LocalSource
from torch._subclasses import FakeTensorMode
from torch.fx.experimental.proxy_tensor import make_fx
from torch.fx.experimental.symbolic_shapes import (
    DimDynamic,
    ShapeEnv,
    StatelessSymbolicContext,
)
from torch.nn.attention import SDPBackend, sdpa_kernel

q = torch.randn(2, 4, 8, 16, device="cuda")
k = torch.randn(2, 4, 8, 16, device="cuda")
v = torch.randn(2, 4, 8, 16, device="cuda")

shape_env = ShapeEnv()
fake_mode = FakeTensorMode(allow_non_fake_inputs=True, shape_env=shape_env)
ctx = StatelessSymbolicContext(
    dynamic_sizes=[
        DimDynamic.UNBACKED,
        DimDynamic.STATIC,
        DimDynamic.STATIC,
        DimDynamic.STATIC,
    ],
    shape_ids={0: "batch"},
    unbacked_bounds={0: (2, 8)},
)

with fake_mode, shape_env.ignore_fresh_unbacked_symbols():
    fq = fake_mode.from_tensor(
        q,
        source=LocalSource("q", is_input=True),
        symbolic_context=ctx,
    )
    fk = fake_mode.from_tensor(
        k,
        source=LocalSource("k", is_input=True),
        symbolic_context=ctx,
    )
    fv = fake_mode.from_tensor(
        v,
        source=LocalSource("v", is_input=True),
        symbolic_context=ctx,
    )

print(fq.shape)  # torch.Size([u0, 4, 8, 16])

def f(q, k, v):
    with sdpa_kernel([SDPBackend.MATH], set_priority=True):
        return F.scaled_dot_product_attention(q, k, v, is_causal=True)

with fake_mode:
    gm = make_fx(f)(fq, fk, fv)
```

## Actual result

Tracing fails with:

```text
torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode:
Could not extract specialized integer from data-dependent expression u0
```

The stack points at fake tensor handling for `aten.size.default`, which returns `int[]` and therefore attempts to convert the unbacked `SymInt` batch size to a concrete Python `int`.

## Expected result

SDPA tracing should not require concrete materialization of the batch size for shape propagation. The SDPA fake/decomposition path should preserve symbolic sizes by using per-dimension symbolic size APIs such as `sym_size.int` or `tensor.shape[i]`/`tensor.size(i)` where a `SymInt` can flow, instead of requesting the full concrete `aten.size.default` tuple.

## TorchTitan impact

The graph trainer chunk passes need to mark selected module input dimensions dynamic before tracing, then compare graph-pass chunking against an eager chunked baseline. For DeepSeek transformer-block granularity, real SDPA currently blocks the trace. The current test workaround is to replace inner attention with a trace-friendly attention-shaped module, which validates chunking mechanics but not real SDPA numerics.

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stacked PRs:
 * #183545
 * #183544
 * #183495
 * __->__#183398

--- --- ---

### [ATen][ProxyTensor] Preserve unbacked batch dims in SDPA tracing

The general batched matmul path still converted broadcast batch dimensions to concrete int64_t sizes before expanding, reshaping, and viewing through bmm. That breaks tracing for rank-4 inputs with an unbacked batch symbol, including the math scaled-dot-product-attention path, because SDPA lowers through batched matmul.

Carry symbolic sizes through the broadcasted bmm path by replacing sizes() and IntArrayRef/DimVector shape plumbing with sym_sizes(), SymDimVector, infer_size_symdimvector(), expand_symint(), reshape_symint(), _unsafe_view_symint(), and symbolic output resize. Keep the existing gradient broadcast shortcut but evaluate its symbolic predicates with guard helpers.

Also preserve symbolic size queries during proxy tracing. aten.size.default has a non-symbolic int[] schema, so executing it on a fake tensor with unbacked sizes forces the symbol and can raise GuardOnDataDependentSymNode. When torch-function metadata tracing sees aten.size.default on a tensor with symbolic sizes, redirect it to aten.sym_size.default so downstream users receive symbolic size proxies instead of forcing concrete integers. This covers cuDNN SDPA fake outputs whose batch dimension remains symbolic.

Add fake tensor regressions for rank-4 matmul, math SDPA, symbolic aten.size.default tracing, and cuDNN SDPA fake output sizes with an unbacked batch dimension.

Test Plan:

python test/test_fake_tensor.py FakeTensorTest.test_linear_rank3_unbacked_batch_dim FakeTensorTest.test_matmul_rank4_unbacked_batch_dim FakeTensorTest.test_sdpa_math_unbacked_batch_dim FakeTensorTest.test_size_default_unbacked_batch_dim FakeTensorTest.test_cudnn_sdpa_unbacked_batch_dim

pytest test/test_fake_tensor.py -k 'linear_rank3_unbacked_batch_dim or matmul_rank4_unbacked_batch_dim or sdpa_math_unbacked_batch_dim or size_default_unbacked_batch_dim or cudnn_sdpa_unbacked_batch_dim' -q

lintrunner --config=.lintrunner.toml aten/src/ATen/native/LinearAlgebra.cpp torch/_decomp/decompositions.py torch/fx/experimental/proxy_tensor.py test/test_fake_tensor.py

Fixes pytorch/torchtitan#3324

### CODE REVIEW

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -2118,51 +2118,66 @@ static Tensor _matmul_impl(
...
-                                                             ret.append({n, m1});
-                                                             return ret; }();
-    const auto tensor1_expanded = tensor1.expand(tensor1_expand_size)
-                                         .reshape({expand_batch_product, n, m1});
+    const auto tensor1_expand_size = [&output_shape, n, m1]{
+      c10::SymDimVector ret(output_shape);
+      ret.append({n, m1});
+      return ret;
+    }();
+    const auto tensor1_expanded = at::reshape_symint(
```
Comment (laithsakka) [inline aten/src/ATen/native/LinearAlgebra.cpp]: nit:
const auto tensor1_expanded = tensor1
    .expand_symint(tensor1_expand_size)
    .reshape_symint({expand_batch_product, n, m1});


better readability same for other locations

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -2118,51 +2118,66 @@ static Tensor _matmul_impl(
...
       } else {
         ret.append({m2, p});
       }
       return ret;
     }();
-    auto tensor2_expanded = tensor2.expand(tensor2_expand_size);
+    auto tensor2_expanded = tensor2.expand_symint(tensor2_expand_size);
     if (vector_rhs) {
-      tensor2_expanded = tensor2_expanded.reshape({expand_batch_product, m2}).unsqueeze(2);
+      tensor2_expanded = at::reshape_symint(
```
Comment (laithsakka) [inline aten/src/ATen/native/LinearAlgebra.cpp]: nit:
tensor2_expanded = tensor2_expanded
    .reshape_symint({expand_batch_product, m2})
    .unsqueeze(2);

**Review by laithsakka — APPROVED** [review]

File: test/test_fake_tensor.py [hunk 3]
```diff
@@ -96,6 +98,26 @@ def expectedFailurePropagateRealTensors(fn):
 
 
 class FakeTensorTest(TestCase):
+    def fake_with_unbacked_batch(self, *tensors):
+        shape_env = ShapeEnv()
+        fake_mode = FakeTensorMode(allow_non_fake_inputs=True, shape_env=shape_env)
+        fakes = []
+        with fake_mode, shape_env.ignore_fresh_unbacked_symbols():
```
[the remark this answers is not in the document]
Comment (sanketpurandare) [reply test/test_fake_tensor.py:106]: Added a comment here explaining why this helper intentionally uses `ignore_fresh_unbacked_symbols()`. The helper creates already-bound fake input batch symbols before tracing; without the context those input symbols remain in `pending_fresh_unbacked_symbols`, and `make_fx` fails while binding the fake inputs instead of exercising the SDPA/matmul behavior under test.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 1 above)
Comment (sanketpurandare) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Updated the `tensor1_expanded` expression to use the chained `expand_symint(...).reshape_symint(...)` form.

File: aten/src/ATen/native/LinearAlgebra.cpp (hunk 2 above)
Comment (sanketpurandare) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Updated the `tensor2_expanded` reshape path to use the chained form before `unsqueeze(2)`.
